File: src/test/java/ai/aimachineserver/domain/games/soccer/BoardSoccerTest.java
package ai.aimachineserver.domain.games.soccer;

import ai.aimachineserver.domain.games.tictactoe.BoardCoords;
import org.junit.jupiter.api.Test;

import java.util.stream.IntStream;

import static ai.aimachineserver.domain.games.soccer.BoardSoccer.*;
import static ai.aimachineserver.domain.games.soccer.NodeLink.*;
import static org.assertj.core.api.Assertions.assertThat;

class BoardSoccerTest {

    @Test
    void getCurrentNode__afterInitialization__currentNodeIsInTheMiddle() {
        BoardSoccer board = new BoardSoccer();
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(new BoardCoords(
                BoardSoccer.middleRowIndex,
                BoardSoccer.middleColIndex));
    }

    @Test
    void makeLink__moveInBounds__currentNodeHasCorrectIndices() {
        BoardSoccer board = new BoardSoccer();
        board.makeLink(LINK_TOP);
        board.makeLink(LINK_RIGHT);
        board.makeLink(LINK_RIGHT);
        int expectedRowIndex = BoardSoccer.middleRowIndex - 1;
        int expectedColIndex = BoardSoccer.middleColIndex + 2;
        BoardCoords expected = new BoardCoords(expectedRowIndex, expectedColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void makeLink__moveInBoundsOtherDirection__currentNodeHasCorrectIndices() {
        BoardSoccer board = new BoardSoccer();
        board.makeLink(LINK_LEFT);
        board.makeLink(LINK_TOP_RIGHT);
        board.makeLink(LINK_TOP_LEFT);
        board.makeLink(LINK_LEFT);
        board.makeLink(LINK_LEFT);
        board.makeLink(LINK_BOTTOM_RIGHT);
        board.makeLink(LINK_BOTTOM);
        board.makeLink(LINK_BOTTOM);
        board.makeLink(LINK_BOTTOM_LEFT);
        int expectedRowIndex = 8;
        int expectedColIndex = 2;
        BoardCoords expected = new BoardCoords(expectedRowIndex, expectedColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void makeLink__tryMoveOutsideBoundsRight__currentNodeStaysInBounds() {
        BoardSoccer board = new BoardSoccer();
        IntStream.rangeClosed(0, 2 * BoardSoccer.BOARD_WIDTH).forEach(
                i -> board.makeLink(LINK_RIGHT)
        );
        int expectedColIndex = BoardSoccer.BOARD_WIDTH - 1;
        BoardCoords expected = new BoardCoords(BoardSoccer.middleRowIndex, expectedColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void makeLink__tryMoveOutsideBoundsLeft__currentNodeStaysInBounds() {
        BoardSoccer board = new BoardSoccer();
        IntStream.rangeClosed(0, 2 * BoardSoccer.BOARD_WIDTH).forEach(
                i -> board.makeLink(LINK_LEFT)
        );
        int expectedColIndex = 1;
        BoardCoords expected = new BoardCoords(BoardSoccer.middleRowIndex, expectedColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void makeLink__tryMoveOutsideBoundsTop__colInsideGate__currentNodeStaysInBounds() {
        BoardSoccer board = new BoardSoccer();
        IntStream.rangeClosed(0, 2 * BoardSoccer.BOARD_HEIGHT).forEach(
                i -> board.makeLink(LINK_TOP)
        );
        int expectedRowIndex = 0;
        BoardCoords expected = new BoardCoords(expectedRowIndex, BoardSoccer.middleColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void makeLink__tryMoveOutsideBoundsDown__colInsideGate__currentNodeStaysInBounds() {
        BoardSoccer board = new BoardSoccer();
        IntStream.rangeClosed(0, BoardSoccer.BOARD_HEIGHT).forEach(
                i -> board.makeLink(LINK_BOTTOM)
        );
        BoardCoords expected = new BoardCoords(BoardSoccer.BOARD_HEIGHT, BoardSoccer.middleColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void makeLink__tryMoveOutsideBoundsTop__colOutsideGate__currentNodeStaysInBounds() {
        BoardSoccer board = new BoardSoccer();
        IntStream.rangeClosed(0, BoardSoccer.GATE_WIDTH).forEach(
                i -> board.makeLink(LINK_RIGHT)
        );
        IntStream.rangeClosed(0, 2 * BoardSoccer.BOARD_HEIGHT).forEach(
                i -> board.makeLink(LINK_TOP)
        );
        int expectedRowIndex = 1;
        int expectedColIndex = BoardSoccer.middleColIndex + GATE_WIDTH + 1;
        BoardCoords expected = new BoardCoords(expectedRowIndex, expectedColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void makeLink__tryMoveOutsideBoundsDown__colOutsideGate__currentNodeStaysInBounds() {
        BoardSoccer board = new BoardSoccer();
        IntStream.rangeClosed(0, GATE_WIDTH).forEach(
                i -> board.makeLink(LINK_LEFT)
        );
        IntStream.rangeClosed(0, 2 * BOARD_HEIGHT).forEach(
                i -> board.makeLink(LINK_BOTTOM)
        );
        int expectedRowIndex = BOARD_HEIGHT - 1;
        int expectedColIndex = BoardSoccer.middleColIndex - GATE_WIDTH - 1;
        BoardCoords expected = new BoardCoords(expectedRowIndex, expectedColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void makeLink__tryMoveOutsideRightBorderAndThenDown__currentNodeStaysInPlace() {
        BoardSoccer board = new BoardSoccer();
        IntStream.rangeClosed(0, 2 * BOARD_WIDTH).forEach(
                i -> board.makeLink(LINK_RIGHT)
        );
        board.makeLink(LINK_BOTTOM);
        int expectedColIndex = BOARD_WIDTH - 1;
        BoardCoords expected = new BoardCoords(BoardSoccer.middleRowIndex, expectedColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void makeLink__tryMoveOutsideBoundsBottomRight__currentNodeStaysInBounds() {
        BoardSoccer board = new BoardSoccer();
        IntStream.range(1, BoardSoccer.middleRowIndex - 1).forEach(
                i -> board.makeLink(LINK_BOTTOM)
        );
        IntStream.range(1, BoardSoccer.middleColIndex - 1).forEach(
                i -> board.makeLink(LINK_RIGHT)
        );
        board.makeLink(LINK_BOTTOM_RIGHT);
        board.makeLink(LINK_BOTTOM_RIGHT);
        int expectedRowIndex = BOARD_HEIGHT - 1;
        int expectedColIndex = BOARD_WIDTH - 1;
        BoardCoords expected = new BoardCoords(expectedRowIndex, expectedColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void makeLink__tryMoveOutsideBoundsBottomSkew__currentNodeStaysInBounds() {
        BoardSoccer board = new BoardSoccer();
        IntStream.range(1, middleRowIndex - 1).forEach(
                i -> board.makeLink(LINK_BOTTOM)
        );
        IntStream.range(1, middleColIndex - 3).forEach(
                i -> board.makeLink(LINK_RIGHT)
        );
        board.makeLink(LINK_BOTTOM_RIGHT);
        board.makeLink(LINK_BOTTOM_LEFT);
        int expectedRowIndex = BOARD_HEIGHT - 1;
        int expectedColIndex = BOARD_WIDTH - 3;
        BoardCoords expected = new BoardCoords(expectedRowIndex, expectedColIndex);
        BoardSoccer.Node node = board.getCurrentNode();
        BoardCoords result = new BoardCoords(node.rowIndex, node.colIndex);
        assertThat(result).isEqualTo(expected);
    }

    @Test
    void isFieldAvailable__topLeftBorder__moveTopRight__returnFalse() {
        BoardSoccer board = new BoardSoccer();
        IntStream.range(0, middleColIndex - 1).forEach(
                i -> board.makeLink(LINK_LEFT)
        );
        IntStream.range(0, middleRowIndex - 1).forEach(
                i -> board.makeLink(LINK_TOP)
        );
        boolean result = board.isFieldAvailable(0, 0);
        assertThat(result).isEqualTo(false);
    }

    @Test
    void isFieldAvailable__bottomRightBorder__moveBottomRight__returnFalse() {
        BoardSoccer board = new BoardSoccer();
        IntStream.range(0, middleColIndex - 1).forEach(
                i -> board.makeLink(LINK_RIGHT)
        );
        IntStream.range(0, middleRowIndex - 1).forEach(
                i -> board.makeLink(LINK_BOTTOM)
        );
        boolean result = board.isFieldAvailable(BOARD_HEIGHT, BOARD_WIDTH);
        assertThat(result).isEqualTo(false);
    }
}

File: src/test/java/ai/aimachineserver/domain/games/soccer/BoardSoccerTest.kt
package ai.aimachineserver.domain.games.soccer

import ai.aimachineserver.domain.games.soccer.BoardSoccer.Companion.BOARD_HEIGHT
import ai.aimachineserver.domain.games.soccer.BoardSoccer.Companion.BOARD_WIDTH
import ai.aimachineserver.domain.games.soccer.BoardSoccer.Companion.GATE_WIDTH
import ai.aimachineserver.domain.games.soccer.BoardSoccer.Companion.middleColIndex
import ai.aimachineserver.domain.games.soccer.BoardSoccer.Companion.middleRowIndex
import ai.aimachineserver.domain.games.soccer.NodeLink.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class BoardSoccerTest {

    @Test
    fun getCurrentNode__afterInitialization__currentNodeIsInTheMiddle() {
        val board = BoardSoccer()
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(middleRowIndex to middleColIndex)
    }

    @Test
    fun makeLink__moveInBounds__currentNodeHasCorrectIndices() {
        val board = BoardSoccer()
        board.makeLink(LINK_TOP)
        board.makeLink(LINK_RIGHT)
        board.makeLink(LINK_RIGHT)
        val expectedRowIndex = middleRowIndex - 1
        val expectedColIndex = middleColIndex + 2
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun makeLink__moveInBoundsOtherDirection__currentNodeHasCorrectIndices() {
        val board = BoardSoccer()
        board.makeLink(LINK_LEFT)
        board.makeLink(LINK_TOP_RIGHT)
        board.makeLink(LINK_TOP_LEFT)
        board.makeLink(LINK_LEFT)
        board.makeLink(LINK_LEFT)
        board.makeLink(LINK_BOTTOM_RIGHT)
        board.makeLink(LINK_BOTTOM)
        board.makeLink(LINK_BOTTOM)
        board.makeLink(LINK_BOTTOM_LEFT)
        val expectedRowIndex = 8
        val expectedColIndex = 2
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun makeLink__tryMoveOutsideBoundsRight__currentNodeStaysInBounds() {
        val board = BoardSoccer()
        (0..(2 * BOARD_WIDTH)).forEach { _ ->
            board.makeLink(LINK_RIGHT)
        }
        val expectedRowIndex = middleRowIndex
        val expectedColIndex = BOARD_WIDTH - 1
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun makeLink__tryMoveOutsideBoundsLeft__currentNodeStaysInBounds() {
        val board = BoardSoccer()
        (0..(2 * BOARD_WIDTH)).forEach { _ ->
            board.makeLink(LINK_LEFT)
        }
        val expectedRowIndex = middleRowIndex
        val expectedColIndex = 1
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun makeLink__tryMoveOutsideBoundsTop__colInsideGate__currentNodeStaysInBounds() {
        val board = BoardSoccer()
        (0..(2 * BOARD_HEIGHT)).forEach { _ ->
            board.makeLink(LINK_TOP)
        }
        val expectedRowIndex = 0
        val expectedColIndex = middleColIndex
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun makeLink__tryMoveOutsideBoundsDown__colInsideGate__currentNodeStaysInBounds() {
        val board = BoardSoccer()
        (0..(2 * BOARD_HEIGHT)).forEach { _ ->
            board.makeLink(LINK_BOTTOM)
        }
        val expectedRowIndex = BOARD_HEIGHT
        val expectedColIndex = middleColIndex
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun makeLink__tryMoveOutsideBoundsTop__colOutsideGate__currentNodeStaysInBounds() {
        val board = BoardSoccer()
        (0..GATE_WIDTH).forEach { _ ->
            board.makeLink(LINK_RIGHT)
        }
        (0..(2 * BOARD_HEIGHT)).forEach { _ ->
            board.makeLink(LINK_TOP)
        }
        val expectedRowIndex = 1
        val expectedColIndex = middleColIndex + GATE_WIDTH + 1
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun makeLink__tryMoveOutsideBoundsDown__colOutsideGate__currentNodeStaysInBounds() {
        val board = BoardSoccer()
        (0..GATE_WIDTH).forEach { _ ->
            board.makeLink(LINK_LEFT)
        }
        (0..(2 * BOARD_HEIGHT)).forEach { _ ->
            board.makeLink(LINK_BOTTOM)
        }
        val expectedRowIndex = BOARD_HEIGHT - 1
        val expectedColIndex = middleColIndex - GATE_WIDTH - 1
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun makeLink__tryMoveOutsideRightBorderAndThenDown__currentNodeStaysInPlace() {
        val board = BoardSoccer()
        (0..(2 * BOARD_WIDTH)).forEach { _ ->
            board.makeLink(LINK_RIGHT)
        }
        board.makeLink(LINK_BOTTOM)
        val expectedRowIndex = middleRowIndex
        val expectedColIndex = BOARD_WIDTH - 1
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun makeLink__tryMoveOutsideBoundsBottomRight__currentNodeStaysInBounds() {
        val board = BoardSoccer()
        (1 until middleRowIndex - 1).forEach { _ ->
            board.makeLink(LINK_BOTTOM)
        }
        (1 until middleColIndex - 1).forEach { _ ->
            board.makeLink(LINK_RIGHT)
        }
        board.makeLink(LINK_BOTTOM_RIGHT)
        board.makeLink(LINK_BOTTOM_RIGHT)
        val expectedRowIndex = BOARD_HEIGHT - 1
        val expectedColIndex = BOARD_WIDTH - 1
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun makeLink__tryMoveOutsideBoundsBottomSkew__currentNodeStaysInBounds() {
        val board = BoardSoccer()
        (1 until middleRowIndex - 1).forEach { _ ->
            board.makeLink(LINK_BOTTOM)
        }
        (1 until BOARD_WIDTH - middleColIndex - 3).forEach { _ ->
            board.makeLink(LINK_RIGHT)
        }
        board.makeLink(LINK_BOTTOM_RIGHT)
        board.makeLink(LINK_BOTTOM_LEFT)
        val expectedRowIndex = BOARD_HEIGHT - 1
        val expectedColIndex = BOARD_WIDTH - 3
        val result = board.getCurrentNode().let { it.rowIndex to it.colIndex }
        assertThat(result).isEqualTo(expectedRowIndex to expectedColIndex)
    }

    @Test
    fun isFieldAvailable__topLeftBorder__moveTopRight__returnFalse() {
        val board = BoardSoccer()
        (0 until middleColIndex - 1).forEach { _ ->
            board.makeLink(LINK_LEFT)
        }
        (0 until middleRowIndex - 1).forEach { _ ->
            board.makeLink(LINK_TOP)
        }
        val result = board.isFieldAvailable(0, 0)
        assertThat(result).isEqualTo(false)
    }

    @Test
    fun isFieldAvailable__bottomRightBorder__moveBottomRight__returnFalse() {
        val board = BoardSoccer()
        (0 until middleColIndex - 1).forEach { _ ->
            board.makeLink(LINK_RIGHT)
        }
        (0 until middleRowIndex - 1).forEach { _ ->
            board.makeLink(LINK_BOTTOM)
        }
        val result = board.isFieldAvailable(BOARD_HEIGHT, BOARD_WIDTH)
        assertThat(result).isEqualTo(false)
    }
}
